misk-hibernate-testing/src/main/kotlin/misk/jdbc/DatabaseAllocator.kt

```diff
@@ -0,0 +1,213 @@
+package misk.jdbc
+
+import misk.hibernate.Session
+import misk.hibernate.Transacter
+import misk.hibernate.shards
+import misk.hibernate.transaction
+import java.time.Clock
+import java.time.Duration
+import javax.inject.Inject
+import javax.inject.Provider
+import javax.inject.Singleton
+import javax.persistence.PersistenceException
+import kotlin.reflect.KClass
+
+/**
+ * This is used to allocate a new database named like `[hint]__[now]__[latestVersion]` using the
+ * base name specified by a [DataSourceConfig].
+ *
+ * [DatabaseAllocator]s are to be used _only_ in tests, where we might want to have create and use
+ * a new database between tests to avoid parallelism issues.
+ *
+ */
+@Singleton
+class DatabaseAllocator(
+  val qualifier: KClass<out Annotation>,
+  val backend: Backend,
+  val clock: Clock,
+  config: DataSourceConfig,
+  retention: Duration = Duration.ofDays(2)
+) {
+  init {
+    backend.type = config.type
```

Comment:
should the backend just get the config? Or should it be a parameter? Mutating it here feels bad

Reply:
Made type a parameter